GetPendingModeratorRequestsPaginatedDone

aabdoo23 opened the pull request:
#### PR Classification
Code refactor and feature enhancement.
Too many changes because it is based on the previous branch not origin
#### PR Summary
Refactored repository classes and added new methods, updated service configurations, and added migrations.
- `BaseUserRepository`: Constructor now throws `ArgumentNullException` if `SchedsForumsDbContext` is null.
- `BaseRepository`: Refactored to use parameterized constructor; added `GetTotalCount` and `GetFromTo` methods.
- `ServiceCollectionExtensions`: Updated repository registrations and refactored JWT configuration methods.
- `SchedsForums.Infrastructure.csproj`: Added new `Migrations` folder.

aabdoo23 committed:
Add feature to fetch pending moderator sign-up requests

Introduced `GetPendingModeratorsQuery` and its handler to fetch
pending moderator sign-up requests with pagination. Added a new
endpoint `GetModeratorSignUpRequestsAsync` in `AdminController`
to handle this query. Created `IPendingModeratorRepository` and
implemented `PendingModeratorRepository` for database operations.
Updated `ModeratorSignUpCommandHandler` to use the new repository.
Added `GetPendingModeratorsQueryResponseDTO` and
`GetPendingModeratorsQueryValidator`. Renamed database table
`ModeratorSignUpRequests` to `PendingModerators`. Updated DI
configuration and added null checks in constructors.

aabdoo23 committed:
Refactor pending moderators handling and repository logic

Replaced GetModeratorSignUpRequests with GetPendingModerators in AdminController.cs.
Extended IBaseRepository.cs with GetFromTo and GetTotalCount methods.
Removed IPendingModeratorRepository.cs and PendingModeratorRepository.cs.
Updated GetPendingModeratorsQueryHandler.cs to use IBaseRepository<PendingModerator>.
Simplified pagination logic and added total count retrieval.
Updated GetPendingModeratorsQueryResponseDTO.cs with TotalCount, PageSize, and PageNumber properties.
Refactored BaseRepository.cs for constructor initialization and added new methods.
Updated ServiceCollectionExtensions.cs to register IBaseRepository<PendingModerator>.

aabdoo23 committed:
Teeny Tiny error

aabdoo23 committed:
merge and update branch via last APPROVED PR

aabdoo23 committed:
small error, fixed, we cool.

Fix JWT_AUDIENCE constant to reference audience config

Updated the JWT_AUDIENCE constant in the JWTConstants class to
reference the audience configuration option instead of the issuer
configuration option. This ensures the correct audience value is
used in the JWT configuration.

aabdoo23 committed:
fixed the pending moderator get retreiving moderators too

Refactor repositories and handlers for PendingModerator

Updated ModeratorSignUpCommandHandler and GetPendingModeratorsQueryHandler to use IPendingModeratorRepository. Modified IBaseRepository to use Guid instead of string for identifiers. Created IPendingModeratorRepository and its implementation PendingModeratorRepository. Updated BaseRepository and ServiceCollectionExtensions accordingly.

MostafaHgz commented on SchedsForums.Api/Controllers/AdminController.cs:
```diff
@@ -19,5 +20,12 @@ public async Task<IActionResult> CreateAdminAsync([FromBody] CreateAdminCommand
         }
+
+        [HttpGet("PendingModerators")]
+        public async Task<IActionResult> GetModeratorSignUpRequestsAsync([FromBody] GetPendingModeratorsQuery query)
```
Thishould be in a new controller

MostafaHgz commented on SchedsForums.Api/appsettings.Development.json:
```diff
@@ -5,8 +5,8 @@
       "Microsoft.AspNetCore": "Warning"
     }
   },
-  "JWT": {
```
Pull from master to update this branch

aabdoo23 replied:
updated

MostafaHgz commented on SchedsForums.Application/Interfaces/Common/IBaseRepository.cs:
```diff
@@ -5,9 +5,12 @@ namespace SchedsForums.Application.Interfaces.Common
         Task<T> UpdateAsync(T entity);
-        Task DeleteAsync(string id);
+        Task DeleteAsync(Guid id);
+        Task<IEnumerable<T>> GetFromTo(int pageNumber, int pageSize);
```
Renamed to get paginated

MostafaHgz commented on SchedsForums.Application/Queries/Admins/GetPendingModerators/GetPendingModeratorsQueryHandler.cs:
```diff
@@ -0,0 +1,30 @@
+            var pendingModeratorsQueryResult = await _pendingModeratorsRepository.GetFromTo(request.PageNumber, request.PageSize);
+            var totalCount = await _pendingModeratorsRepository.GetTotalCount();
+
+            return new GetPendingModeratorsQueryResponseDTO
```
Create a single paginted bas response

MostafaHgz replied:
Paginated response should return from the query not here as it's common in all queries

aabdoo23 committed:
Comments fixed, only waiting for the last comment explanation to fix all, pushed to work on another branch

Refactor pending moderators handling and pagination

- Remove `GetPendingModerators` from `AdminController.cs`
- Introduce `ModeratorRequestsController` for pending moderators
- Rename `GetFromTo` to `GetPaginated` in `IBaseRepository`
- Update repositories to use `GetPaginated` method
- Replace `BaseGetMultipleQuery` with `BaseGetPaginatedQuery`
- Add `BaseGetPaginatedResponseDTO` for paginated responses
- Introduce `BaseGetPaginatedValidator` for query validation
- Move and update pending moderators query classes to new namespace

aabdoo23 committed:
dun da dun dun

Add GetPaginatedAsync method and update DTOs

- Added GetPaginatedAsync method to IBaseRepository interface.
- Changed BaseGetPaginatedResponseDTO from abstract to concrete.
- Renamed Content property to Data in BaseGetPaginatedResponseDTO.
- Updated GetPendingModeratorsQueryHandler to use GetPaginatedAsync.
- Implemented GetPaginatedAsync in BaseRepository to handle pagination.

aabdoo23 committed:
hella lot of abstractions

Refactor DTOs and update pagination methods

Updated `Id` property type from `string` to `Guid` in various command handlers and response DTOs. Introduced `BaseResponseDTO` and modified `UserResponseDTO` to inherit from it. Refactored pagination methods in `IBaseRepository` and `IPendingModeratorRepository` interfaces, adding `GetPaginatedContentAsync` and `GetPaginatedPendingModeratorsAsync`. Removed `BaseGetPaginatedResponseDTO.cs` and replaced its functionality with new DTO classes. Updated repository and query handler classes to use new DTOs and methods. Added new DTO classes for consistent response handling.

MostafaHgz commented on SchedsForums.Application/Interfaces/Common/DTOs/BaseResponseDTO.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace SchedsForums.Application.Interfaces.Common.DTOs
+{
+    public class BaseResponseDTO
```
rename to BaseIdResponseDTO

MostafaHgz commented on SchedsForums.Application/Queries/Common/DTOs/BaseGetPaginatedDTO.cs:
```diff
@@ -0,0 +1,11 @@
+﻿namespace SchedsForums.Application.Queries.Common.DTOs
+{
+    public abstract class BaseGetPaginatedDTO<T>
```
rename to PaginatedResponseDTO

MostafaHgz commented on SchedsForums.Application/Queries/Common/DTOs/BaseGetPaginatedDTO.cs:
```diff
@@ -0,0 +1,11 @@
+        public int PageNumber { get; set; }
+        public int PageSize { get; set; }
+        public int TotalCount { get; set; }
+        public int ReturnedCount { get; set; }
```
no need for this

MostafaHgz commented on SchedsForums.Application/Queries/Common/DTOs/BaseGetPaginatedDTO.cs:
```diff
@@ -0,0 +1,11 @@
+        public int PageSize { get; set; }
+        public int TotalCount { get; set; }
+        public int ReturnedCount { get; set; }
+        public IEnumerable<T> Data { get; set; }
```
this should be IList<>

MostafaHgz commented on SchedsForums.Application/Queries/Common/DTOs/BaseGetPaginatedEntityDTO.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using SchedsForums.Domain.Entities.Common;
+
+namespace SchedsForums.Application.Queries.Common.DTOs
```
what does this do

MostafaHgz replied:
this should be in the repository

aabdoo23 committed:
Merge branch 'master' into GetPendingModerators

MostafaHgz commented on SchedsForums.Infrastructure/Repositories/PendingModeratorRepository.cs:
```diff
@@ -0,0 +1,38 @@
+            };
+        }
+
+        public override async Task<IEnumerable<PendingModerator>> GetAllAsync()
```
remove this

aabdoo23 committed:
final final final

Refactor DTO inheritance and update pagination methods

Refactored DTO classes to improve inheritance structure:
- `UserResponseDTO` and `BasePendingModeratorResponseDTO` now inherit from `BaseIdResponseDTO`.
- Removed `BaseResponseDTO` and related paginated DTO classes.
- Added `BaseIdResponseDTO`, `PaginatedEntityDTO<T>`, and `PaginatedResponseDTO<T>`.

Updated repository methods to use new paginated DTOs:
- `GetPaginatedContentAsync` in `IBaseRepository` and `BaseRepository` now return `PaginatedEntityDTO<T>`.
- `GetPaginatedPendingModeratorsAsync` in `IPendingModeratorRepository` and `PendingModeratorRepository` now return `PaginatedResponseDTO<BasePendingModeratorResponseDTO>`.
- `GetPendingModeratorsQueryResponseDTO` now inherits from `PaginatedResponseDTO<BasePendingModeratorResponseDTO>`.

MostafaHgz commented on SchedsForums.Application/Queries/Common/DTOs/PaginatedEntityDTO.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using SchedsForums.Domain.Entities.Common;
+
+namespace SchedsForums.Application.Queries.Common.DTOs
```
I believe we agreed this should be moved to the infrastructure

MostafaHgz commented on SchedsForums.Infrastructure/Repositories/Common/BaseRepository.cs:
```diff
@@ -36,11 +33,36 @@ public async Task<IEnumerable<T>> GetAllAsync()
+                .Skip((pageNumber - 1) * pageSize)
+                .Take(pageSize)
+                .ToListAsync();
+            var returnedCount = data.Count;
```
this is not used anymore

aabdoo23 committed:
minor fix for PR comment

Remove unused returnedCount calculation in BaseRepository

The line `var returnedCount = data.Count;` has been removed from the `BaseRepository.cs` file within the `SchedsForums.Infrastructure.Repositories.Common` namespace. This line was responsible for storing the count of the items in the `data` list after the pagination logic was executed. The rest of the code remains unchanged, including the logic for ordering, skipping, and taking a specific number of items from the `queryable` collection, and returning a `PaginatedEntityDTO<T>` object.
